File: leetcode/js/32. Longest Valid Parentheses.js
var longestValidParentheses = function (s) {
  let maxans = 0;
  let stack = [];
  stack.push(-1);
  for (let i = 0; i < s.length; i++) {
    if (s[i] == "(") stack.push(i);
    else {
      stack.pop();
      if (stack.length == 0) {
        stack.push(i);
      } else {
        maxans = Math.max(maxans, i - stack.slice(-1)[0]);
      }
    }
  }
  return maxans;
};

console.log(longestValidParentheses("(()()"));

File: leetcode/kotlin/32. Longest Valid Parentheses.kt
class Solution {

    fun longestValidParentheses(s:String):Int {
       var maxans = 0
        val stack = Stack<Int>()
        stack.push(-1)
        for (i in 0 until s.length) {
            if (s[i] == '(') {
                stack.push(i)
            } else {
                stack.pop()
                if (stack.empty()) {
                    stack.push(i)
                } else {
                    maxans = Math.max(maxans, i - stack.peek())
                }
            }
        }
        return maxans
    }
}